File: 백준/Gold/6443. 애너그램/애너그램.java
import java.util.Arrays;
import java.util.Scanner;

public class Main {
    static int len;
    static boolean visited[];
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextLine().charAt(0) - '0';
        for (int i = 0; i < n; i++) {
            String temp = sc.nextLine();
            char[] s = temp.toCharArray();
            len = s.length;
            Arrays.sort(s);
            visited = new boolean[len];
            dfs(s, new StringBuilder(),0);
        }

    }


    public static void dfs(char[] arr, StringBuilder sb, int depth) {
        if (depth == len) {
            System.out.println(sb.toString());
            return;
        }

        for (int i = 0; i < len; i++) {
            if (visited[i]) continue;
            if (i > 0 && arr[i] == arr[i - 1] && !visited[i - 1]) continue;

            visited[i] = true;
            sb.append(arr[i]);
            dfs(arr,  sb, depth + 1);
            sb.deleteCharAt(sb.length() - 1);
            visited[i] = false;
        }
    }

}

File: 백준/Gold/6443. 애너그램/애너그램.cc
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;
string str;
int len;

void dfs(string s, int idx) {
	if (idx == len - 1) {
		cout << s << '\n';
		return;
	}
	for (int i = idx; i < len; i++) {
		if (i != idx && s[i] == s[idx])
			continue;
		if(s[i] != s[idx]){
			char c = s[i];
			s[i] = s[idx];
			s[idx] = c;
		}
		dfs(s, idx + 1);
	}
}

int main() {
	int n;
	cin >> n;
	while (n) {
		cin >> str;
		len = str.size();
		sort(str.begin(), str.end());
		dfs(str, 0);
		n--;
	}
	return 0;
}
